File: deck/seed/code/native/kotlin/file/runtime/watch.kt
// Filesystem watching for the kotlin target, over java.nio.file.WatchService: the JDK's own watcher, not a
// polling loop and not a dependency.
//
// The API is PULL, not callback: `watchOpen` registers the watch, `watchNext` awaits the next change, and
// `watchClose` stops it. WatchService is already pull-shaped (`take()` blocks until a change), so the whole shim
// is that call moved onto `Dispatchers.IO` and the JDK's three event kinds renamed to the four the Term API uses.
//
// Reached only through the public file/watch API.
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.StandardWatchEventKinds
import java.nio.file.WatchService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object watch {
  // the running watch: the service, the directories registered with it (so a key can be turned back into a path),
  // and whether it has been closed
  class Watcher(val service: WatchService?, val roots: MutableMap<Any, Path>) {
    var closed = false
  }

  suspend fun watchOpen(path: String, deep: Boolean): Watcher = withContext(Dispatchers.IO) {
    try {
      val service = FileSystems.getDefault().newWatchService()
      val roots = mutableMapOf<Any, Path>()
      val root = Paths.get(path)

      register(service, roots, root)

      if (deep) {
        // WatchService registers ONE directory at a time, so a deep watch is every directory beneath the root
        // registered up front. A directory created later is not covered, which is the JDK's own limitation and
        // is named in note/term/stdlib/native-async-file-and-server.md.
        Files.walk(root).use { tree ->
          for (child in tree) {
            if (Files.isDirectory(child) && child != root) {
              register(service, roots, child)
            }
          }
        }
      }

      Watcher(service, roots)
    } catch (error: Exception) {
      Watcher(null, mutableMapOf())
    }
  }

  private fun register(service: WatchService, roots: MutableMap<Any, Path>, at: Path) {
    val key =
      at.register(
        service,
        StandardWatchEventKinds.ENTRY_CREATE,
        StandardWatchEventKinds.ENTRY_DELETE,
        StandardWatchEventKinds.ENTRY_MODIFY,
      )
    roots[key] = at
  }

  // the next change. A closed watcher answers with the empty event rather than waiting forever, so a loop over
  // `watchNext` ends after `watchClose`.
  suspend fun watchNext(watcher: Watcher): WatchEvent = withContext(Dispatchers.IO) {
    val service = watcher.service

    if (service == null || watcher.closed) {
      return@withContext WatchEvent("", "")
    }

    try {
      while (true) {
        val key = service.take()
        val root = watcher.roots[key] ?: Paths.get(".")
        val events = key.pollEvents()
        key.reset()

        for (event in events) {
          val kind =
            when (event.kind()) {
              StandardWatchEventKinds.ENTRY_CREATE -> "create"
              StandardWatchEventKinds.ENTRY_DELETE -> "remove"
              StandardWatchEventKinds.ENTRY_MODIFY -> "change"
              else -> "other"
            }
          val at = event.context()

          return@withContext WatchEvent(
            kind,
            if (at is Path) root.resolve(at).toString() else root.toString(),
          )
        }
      }

      @Suppress("UNREACHABLE_CODE")
      WatchEvent("", "")
    } catch (error: Exception) {
      WatchEvent("", "")
    }
  }

  suspend fun watchClose(watcher: Watcher) {
    withContext(Dispatchers.IO) {
      watcher.closed = true

      try {
        watcher.service?.close()
      } catch (error: Exception) {
        null
      }
    }
  }
}

File: deck/seed/code/native/rust/file/runtime/watch.rs
// Filesystem watching for the rust target, over the `notify` crate -- the ecosystem-standard watcher, which uses
// FSEvents on macOS, inotify on Linux and ReadDirectoryChangesW on Windows rather than polling.
//
// The API is PULL, not callback: `watch_open` starts the watcher, `watch_next` awaits the next change, and
// `watch_close` stops it. That is the one shape all four backends can hold (node's fs.watch async iterator, this,
// Swift's DispatchSource, the JVM's WatchService), so the Term API above it is the same everywhere. A callback
// shape is not: it would need a `task` value crossing into a platform thread on three of the four.
//
// Reached only through the public file/watch API.
mod watch {
    use super::WatchEvent;
    use std::sync::Arc;
    use tokio::sync::Mutex;

    // the running watcher: the notify handle is held only to keep it alive (dropping it stops the watch), and the
    // receiver is where changes arrive
    pub struct WatchState {
        pub keep: Option<notify::RecommendedWatcher>,
        pub take: tokio::sync::mpsc::UnboundedReceiver<WatchEvent>,
    }

    pub type Watcher = Arc<Mutex<WatchState>>;

    fn name(kind: &notify::EventKind) -> &'static str {
        match kind {
            notify::EventKind::Create(_) => "create",
            notify::EventKind::Remove(_) => "remove",
            notify::EventKind::Modify(_) => "change",
            _ => "other",
        }
    }

    pub async fn watch_open(path: String, deep: bool) -> Watcher {
        use notify::Watcher as _;

        let (send, take) = tokio::sync::mpsc::unbounded_channel();
        let watcher = notify::recommended_watcher(
            move |event: notify::Result<notify::Event>| {
                let event = match event {
                    Ok(event) => event,
                    Err(_) => return,
                };

                for at in event.paths {
                    let _ = send.send(WatchEvent {
                        kind: name(&event.kind).to_string(),
                        path: at.to_string_lossy().to_string(),
                    });
                }
            },
        );

        let keep = match watcher {
            Ok(mut watcher) => {
                let mode = if deep {
                    notify::RecursiveMode::Recursive
                } else {
                    notify::RecursiveMode::NonRecursive
                };
                let _ = watcher.watch(std::path::Path::new(&path), mode);

                Some(watcher)
            }
            Err(_) => None,
        };

        Arc::new(Mutex::new(WatchState { keep, take }))
    }

    // the next change. A closed watcher answers with the empty event rather than blocking forever, so a caller
    // that loops on `watch_next` terminates after `watch_close`.
    pub async fn watch_next(watcher: Watcher) -> WatchEvent {
        match watcher.lock().await.take.recv().await {
            Some(event) => event,
            None => WatchEvent {
                kind: String::new(),
                path: String::new(),
            },
        }
    }

    pub async fn watch_close(watcher: Watcher) {
        let mut state = watcher.lock().await;
        state.keep = None;
        state.take.close();
    }
}
